File: 2019-01-18/2750/2750.java
import java.io.*;

class Main{
	static int [] arr = new int[1003];
	public static void main(String[] args) throws Exception{
		BufferedReader bfrReader = new BufferedReader(new InputStreamReader(System.in));
		int N = strToInt(bfrReader.readLine());
		input(N, bfrReader);
		mergeSort(0, N-1);
		output(N);
	}

	public static void output(int N){
		for(int i = 0; i < N; i++)
			System.out.println(arr[i]);
	}

	public static void mergeSort(int start,int end){
		if (start < end){
			int mid = (start+end)/2;
			mergeSort(start, mid);
			mergeSort(mid+1, end);
			merge(start, mid, end);
		}
	}

	public static void merge(int start, int mid, int end){
		int part1 = start, part2 = mid+1;
		int index = start; 
		int []temp = new int[1003];
		for(int i = start; i <= end; i++)
			temp[i] = arr[i];

		while(part1 <= mid && part2 <= end ){
			if(temp[part1] < temp[part2]){
				arr[index] = temp[part1];
				part1++;
			}else{
				arr[index] = temp[part2];
				part2++;
			}
			index++;
		}

		for(int i = 0; i <= mid-part1; i++)
			arr[i+index] = temp[part1+i];
	}

	public static void input(int N, BufferedReader bfrReader) throws Exception{
		for(int i =0; i < N; i++)
			arr[i] = strToInt(bfrReader.readLine());
	}

	public static int strToInt(String str){
		return Integer.parseInt(str);
	}
}

File: 2019-01-18/2750/2750.c
#include <stdio.h>

void quickSort(int start, int end);
int partition(int start, int end);
void swap(int start, int end);

int arr[1002];

int main(){
	int N;
	
	scanf("%d", &N);

	for(int i = 0; i < N; i++)
		scanf("%d", &arr[i]);

	quickSort(0, N-1);

	for(int i =0; i <N;i++)
		printf("%d\n", arr[i]);
	
}

void quickSort(int start, int end){
	int part = partition(start, end);
	if(start < part-1)
		quickSort(start, part-1);
	if(part<end)
		quickSort(part, end);
}

int partition(int start, int end){
	int pivot = arr[(start+end)/2];

	while(start<=end){
		while(arr[start] < pivot) start++;
		while(arr[end] > pivot) end--;
		if(start <= end){
			swap(start, end);
			start++;
			end--;
		}
	}
	return start;
}

void swap(int start, int end){
	int temp = arr[start];
	arr[start] = arr[end];
	arr[end] = temp;
}